File: 1. Arrays And String/Question-2/Solution.py
#!/usr/bin/env python3

def is_permutation_by_sorting(s, t):
    """
    Solution 1: Sort both strings and compare them
    Time Complexity: O(n log n) due to sorting
    Space Complexity: O(n) for creating sorted copies
    """
    # Check if strings have different lengths
    if len(s) != len(t):
        return False
    
    # Sort and compare
    return sorted(s) == sorted(t)

def is_permutation_by_count(s, t):
    """
    Solution 2: Count character occurrences
    Time Complexity: O(n) where n is the length of the strings
    Space Complexity: O(c) where c is the character set size (constant)
    """
    # Check if strings have different lengths
    if len(s) != len(t):
        return False
    
    # Use a dictionary to track character counts
    char_count = {}
    
    # Count occurrences in first string
    for char in s:
        if char in char_count:
            char_count[char] += 1
        else:
            char_count[char] = 1
    
    # Decrement counts for second string
    for char in t:
        if char not in char_count or char_count[char] == 0:
            return False
        char_count[char] -= 1
    
    return True

def is_permutation_by_count_alternate(s, t):
    """
    Solution 3: Use Python's Counter for character counting
    Time Complexity: O(n)
    Space Complexity: O(c) where c is the character set size
    """
    from collections import Counter
    return Counter(s) == Counter(t)

if __name__ == "__main__":
    pairs = [("apple", "papel"), ("carrot", "tarroc"), ("hello", "llloh")]
    
    print("Using sorting method:")
    for word1, word2 in pairs:
        anagram = is_permutation_by_sorting(word1, word2)
        print(f"{word1}, {word2}: {anagram}")
    
    print("\nUsing character count method:")
    for word1, word2 in pairs:
        anagram = is_permutation_by_count(word1, word2)
        print(f"{word1}, {word2}: {anagram}")
    
    print("\nUsing Counter method:")
    for word1, word2 in pairs:
        anagram = is_permutation_by_count_alternate(word1, word2)
        print(f"{word1}, {word2}: {anagram}")

File: 1. Arrays And String/Question-2/Solution.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
using namespace std;

/**
 * Solution 1: Using sorting approach
 * Time Complexity: O(n log n) due to sorting
 * Space Complexity: O(n) for creating sorted copies
 */
bool isPermutationUsingSorting(string s, string t) {
    // Check if strings have different lengths
    if (s.length() != t.length()) {
        return false;
    }
    
    // Sort both strings
    sort(s.begin(), s.end());
    sort(t.begin(), t.end());
    
    // Compare the sorted strings
    return s == t;
}

/**
 * Solution 2: Using character count
 * Time Complexity: O(n) where n is the length of the strings
 * Space Complexity: O(1) as we use a fixed-size array for character counts
 */
bool isPermutationUsingCharCount(const string& s, const string& t) {
    // Check if strings have different lengths
    if (s.length() != t.length()) {
        return false;
    }
    
    // Create an array to store character counts
    int letters[128] = {0}; // Assumption: ASCII
    
    // Count occurrences in first string
    for (int i = 0; i < s.length(); i++) {
        letters[s[i]]++;
    }
    
    // Decrement counts for second string
    for (int i = 0; i < t.length(); i++) {
        letters[t[i]]--;
        if (letters[t[i]] < 0) {
            return false;
        }
    }
    
    return true;
}

int main() {
    vector<pair<string, string>> pairs = {
        {"apple", "papel"}, 
        {"carrot", "tarroc"}, 
        {"hello", "llloh"}
    };
    
    cout << "Using sorting method:" << endl;
    for (const auto& pair : pairs) {
        const string& word1 = pair.first;
        const string& word2 = pair.second;
        bool anagram = isPermutationUsingSorting(word1, word2);
        cout << word1 << ", " << word2 << ": " << (anagram ? "true" : "false") << endl;
    }
    
    cout << "\nUsing character count method:" << endl;
    for (const auto& pair : pairs) {
        const string& word1 = pair.first;
        const string& word2 = pair.second;
        bool anagram = isPermutationUsingCharCount(word1, word2);
        cout << word1 << ", " << word2 << ": " << (anagram ? "true" : "false") << endl;
    }
    
    return 0;
}
